File: app/hooks/commands/command.js
// Indicates whether a command is a request or a response
export const CommandType = {
  REQUEST: "0",
  RESPONSE: "1",
};

// Enumerates the available commands
export const Command = {
  HEALTHCHECK: 0,
  GET_PEDESTALS: 1,
  SET_PEDESTALS_COLOR: 2,
};

// Converts an integer into a 2 character hex string, mod 256
export const intToAsciiHexByte = (i) => {
  return ("00" + i.toString(16)).slice(-2);
};

// Manages pending and removing command response callbacks
export const commandReducer = (state, action) => {
  switch (action.type) {
    case "pend_request":
      const newState = {
        ...state,
      };
      newState[action.id] = action.callback;
      return newState;
    case "clear_request":
      return Object.fromEntries(
        Object.entries(state).filter(([key]) => key !== action.id)
      );
  }
};

// Parses an ID-command-data packet
export const parseCommand = (raw) => {
  const rawSplit = raw.split("|");
  const rawID = rawSplit[0];
  const id = rawID.slice(1);
  const commandType = rawID[0];
  const commandID = rawSplit[1];
  const command = parseInt(commandID, 16);
  const rawData = rawSplit[2];
  let data;
  if (!rawData.includes("#")) {
    data = rawData.length > 0 ? [rawData] : [];
  } else {
    data = rawData.split("#");
  }
  return [commandType, id, command, data];
};

File: circuit_python/command.py
class CommandType:
    REQUEST = "0"
    RESPONSE = "1"


class Command:
    HEALTHCHECK = 0
    GET_PEDESTALS = 1
    SET_PEDESTALS_COLOR = 2
    BLINK_PEDESTAL = 3


def int_to_ascii_byte(i):
    """Converts an integer into a 2 character hex string, mod 256"""
    return ("00" + "{0:x}".format(i))[-2:]


def parse_command(raw):
    """Parses an ID-command-data packet"""
    raw_split = raw.split("|")
    raw_id, command_id, raw_data = raw_split[0], raw_split[1], raw_split[2]
    command_type = raw_id[0]
    id = raw_id[1:]
    command = int(command_id, 16)
    if not "#" in raw_data:
        data = [raw_data] if len(raw_data) > 0 else []
    else:
        data = raw_data.split("#")

    return command_type, id, command, data
